How the Review Process work (#120)
falcosecurity/.github: contributing/review-process.md

- When expressing your code style preferences, please prioritize specific style modifications that bring clear benefits to the project in terms of performance or maintainability. It is important to avoid excessive emphasis on personal style. This approach promotes inclusivity and respect for different coding styles, as long as they align with the project's best interests.
- Please remember that in the realm of security, striking a balance between speed, caution, and code reorganization is crucial, considering that time is of the essence. This balanced approach allows us to effectively support the necessary capabilities for enhanced threat detection, performance, and scalability benefits within reasonable timelines.

## How the Review Process Works

>  The author submits a PR. 

Reviewer: ```suggestion
 - The author submits a PR.

```

> - Automation(@poiana) suggests reviewers for the PR 

Reviewer: ```suggestion
- Automation ([@poiana](https://github.com/poiana)) suggests reviewers for the PR:
```

>   - Determine the set of OWNERS files.

Reviewer: ```suggestion
  - Determine the set of [OWNERS](https://github.com/falcosecurity/evolution/blob/main/GOVERNANCE.md#repository-ownership) files.
```

>   - choose at least two suggested *reviewers*, trying to find a unique reviewer for every leaf OWNERS file, and request their reviews on the PR. 

Reviewer: ```suggestion
  - Choose at least two suggested *reviewers*, trying to find a unique reviewer for every leaf OWNERS file, and request their reviews on the PR. 
```

> - Humans review the PR 

Reviewer: ```suggestion
- Humans review the PR:
```

>    - Reviewers look for general code quality, correctness, sane software engineering, style, etc.

Reviewer: ```suggestion
   - Reviewers look for general code quality, correctness, style and follow [reviewers guideline](#guidelines-for-code-reviewers)
```

   - Anyone member of falco can act as a reviewer with the exception of the individual who opened the PR.
   - if the code changes look good to them, a reviewer types ```lgtm``` in a PR comment or review; if they change their mind they ```/lgtm cancel``` 

>    - once a reviewer has ```/lgtm```'ed, @poiana applies an ```lgtm``` label to the PR.

Reviewer: ```suggestion
   - Anyone in the determined set of OWNERS (including both `reviewers` and `approvers`) can act as a reviewer, with the exception of the individual who opened the PR.
   - If the code changes look good to them, a reviewer types `/lgtm` in a PR comment or selects the _Approved_ checkbox and clicks the _Submit review_ button on the GitHub UI; if they change their mind, they can use `/lgtm cancel.`
   - Once a reviewer has `/lgtm`'ed, [@poiana](https://github.com/poiana) applies an `lgtm` label to the PR.
```

> - Humans approve the PR 

Reviewer: ```suggestion
- Humans approve the PR:
```

>   - The PR author ```/assign```'s suggested approvers to the PR, and optionally notifies them(eg: pinging @foo for approval)
>   - Only people listed in the relevant OWNERS files, either directly or through an alias, can act as APPROVERS, including the individual who opened the PR. 
>   - Approvers look for holistic acceptance criteria, including dependencies with other features, forwards/backwards compatibility, API and flag definitions, etc
>   - for ``falco`` and ``libs``
>   repositeries we require double approval from maintainers and we should explain why and when using this strategy. 
>   - If the code changes look good to approver, an approver types ``/approve`` in a PR comment; if they change their mind, they ``/approve`` cancel
>   - Alongside the Prow workflow, Falco enforce GitHub reviews too. Simply using ``/approve`` is insufficient; approver must additionally select the 'Approved' checkbox and click the 'Submit review' button on the GitHub UI
>   - A positive GitHub review implies the approved label, so approver doesn't have to write ``/approve`` explicitly; however, ``test-infra`` repositery requires an explicit /approve
>   - Once all approver (atleast one) have approved, @poianna applies an ``approved`` label. 

Reviewer: ```suggestion
  - The PR author may `/assign` suggested approvers to the PR and optionally notify them (eg: pinging @someone for approval)
  - Only [Maintainers](https://github.com/falcosecurity/evolution/blob/main/GOVERNANCE.md#maintainers) (ie. people listed in the determined set of OWNERS `approvers` section) can act as **approvers** (including the individual who opened the PR; however, their self-approval is insufficient to merge the PR). 
  - Approvers look for holistic acceptance criteria, including dependencies with other features, forward/backward compatibility, API and flag definitions, etc
  -  If the code changes look good, an approver must select the _Approved_ checkbox and click the _Submit review_ button on the GitHub UI.
     - This is required since The Falco Project enforces GitHub reviews alongside the Prow workflow.
     - Furthermore, some repositories may require two (or more) positive GitHub reviews (notably https://github.com/falcosecurity/falco and https://github.com/falcosecurity/libs).
  - A positive GitHub review implies the PR is approved and looks good, so the approver doesn't have to type `/approve` or `/lgtm` explicitly (however some repositories - notably https://github.com/falcosecurity/test-infra may require an explicit `/approve`).
  - Once at least one approver has approved, [@poiana](https://github.com/poiana) applies the `approved` label.
```

N.B. The multiple approval requirements are only enforced by GitHub. Prow has no mechanism to count approvers. For this reason we always require to use the GitHub UI, otherwise the PR can't be merged.

> - Automation merges the PR: 
>   - if all of the following are TRUE:
>     - All required labels are present (eg: lgtm, approved)
>     - any blocking labels are missing  (eg: there is no do-not-merge/hold, needs-rebase)
>   - And if any of the following are true:
>     - there are no presubmit prow jobs configured for this repo 
>     - there are presubmit prow jobs configured for this repo, and they all pass after automatically being re-run one last time 
>   - then PR will automatically be merged. 

Reviewer: ```suggestion
- Automation merges the PR: 
  - If all of the following are **true**:
    - All required labels are present (eg: `lgtm`, `approved`);
    - Any blocking labels are missing  (eg: there is no `do-not-merge/hold`, `needs-rebase`):
    - All GitHub PR checks marked as _required_ are passing;
    - The PR got sufficient approving GitHub review(s) (usually, only one positive review is sufficient).
 - And if any of the following are true:
    - There are no presubmit Prow jobs configured for this repo; 
    - There are presubmit Prow jobs configured for this repo, and they all pass after automatically being re-run one last time. 
  - Then PR will automatically be merged by [@poiana](https://github.com/poiana).
```

- Please remember that in the realm of security, striking a balance between speed, caution, and code reorganization is crucial, considering that time is of the essence. This balanced approach allows us to effectively support the necessary capabilities for enhanced threat detection, performance, and scalability benefits within reasonable timelines.

## How the Review Process Works

 - The author submits a PR.

- Automation ([@poiana](https://github.com/poiana)) suggests reviewers for the PR:

  - Determine the set of [OWNERS](https://github.com/falcosecurity/evolution/blob/main/GOVERNANCE.md#repository-ownership) files.
  - Choose at least two suggested *reviewers*, trying to find a unique reviewer for every leaf OWNERS file, and request their reviews on the PR. 
- Humans review the PR:
   - Reviewers look for general code quality, correctness, style and follow [reviewers guideline](#guidelines-for-code-reviewers)
   - Anyone in the determined set of OWNERS (including both `reviewers` and `approvers`) can act as a reviewer, with the exception of the individual who opened the PR.

>    - If the code changes look good to them, a reviewer types `/lgtm` in a PR comment or selects the _Approved_ checkbox and clicks the _Submit review_ button on the GitHub UI; if they change their mind, they can use `/lgtm cancel.`

Reviewer 2: I believe we always need the git approve. It's not an or.

In addition plz change to  "either `/lgtm` or `/approve`" etc.

Author: Thanks @incertum for your review. Is it necessary for reviewers to also click the 'Submit review' button from the GitHub UI, or is '/lgtm' sufficient? I think selecting checkbox is must for approving changes only not for reviewing🤔.

Reviewer 2: I see we want to separate the review from the actual approval. I think the review section then really can just be reviewing the PR and all of the lgtm and approve could be moved to the approval section? @leogr ?

Re approval: You need to check mark the git approve button and click the 'Submit review' button alongside either /lgtm or /approve.

Reviewer 2: <img width="698" alt="image" src="https://github.com/falcosecurity/.github/assets/24579375/651b7ae9-cca0-4204-b9d8-7b8074eb5e65">

Reviewer: > I see we want to separate the review from the actual approval. I think the review section then really can just be reviewing the PR and all of the lgtm and approve could be moved to the approval section? @leogr ?

The process is already separated because [Reviewers](https://github.com/falcosecurity/evolution/blob/main/GOVERNANCE.md#reviewers) are allowed to `/lgtm` but they aren't allowed to `/approve`. On the other hand, *[Maintainers](https://github.com/falcosecurity/evolution/blob/main/GOVERNANCE.md#maintainers)* can both `/lgtm` and `/approve`.

The following statement is correct and works in both cases:
> If the code changes look good to them, a reviewer types `/lgtm` in a PR comment or selects the _Approved_ checkbox and clicks the _Submit review_ button on the GitHub UI; if they change their mind, they can use `/lgtm cancel.`

Explanation: 
- When a Reviewer (ie. a user listed under `reviewers` in the OWNERS) approves from GitHub UI, only `lgtm` is automatically added.
- When a Maintainer (ie. a user listed under `approvers` in the OWNERS) approves from GitHub UI, both `lgtm` and `approved` are automatically added.

Typing `/approve` is never required (the only exception is test-infra). Moreover, reviewers must not use `/approve`.

However, since using the GitHub UI is completely optional for reviewers, if we want to make it a bit simpler here and not mix the reviewer and approver cases, we can just omit the GitHub UI part and say that reviewers must only type `/lgtm`. I'll leave a suggestion for this.

>    - If the code changes look good to them, a reviewer types `/lgtm` in a PR comment or selects the _Approved_ checkbox and clicks the _Submit review_ button on the GitHub UI; if they change their mind, they can use `/lgtm cancel.`

Reviewer: ```suggestion
   - If the code changes look good to them, a reviewer types `/lgtm` in a PR comment; if they change their mind, they can use `/lgtm cancel.`
```

>    - Once a reviewer has `/lgtm`'ed, [@poiana](https://github.com/poiana) applies an `lgtm` label to the PR.

Reviewer 2: I would just remove these bullet points wrt lgtm as we outline this in the following section.

- Humans approve the PR:
  - The PR author may `/assign` suggested approvers to the PR and optionally notify them (eg: pinging @someone for approval)
  - Only [Maintainers](https://github.com/falcosecurity/evolution/blob/main/GOVERNANCE.md#maintainers) (ie. people listed in the determined set of OWNERS `approvers` section) can act as **approvers** (including the individual who opened the PR; however, their self-approval is insufficient to merge the PR). 
  - Approvers look for holistic acceptance criteria, including dependencies with other features, forward/backward compatibility, API and flag definitions, etc

>   -  If the code changes look good, an approver must select the _Approved_ checkbox and click the _Submit review_ button on the GitHub UI.

Reviewer 2: here its is the correct way of saying `must select the _Approved_ checkbox`

Author: Yeah, i guess it sounds different. Should we change it to `an approver needs to select`?

>   -  If the code changes look good, an approver must select the _Approved_ checkbox and click the _Submit review_ button on the GitHub UI.

Reviewer 2: ```suggestion
  -  If the code changes look good, the approver navigates to the `Reviews changes` section in the GitHub UI. They then enter either `/lgtm` or `/approve` as a message, check the GitHub `Approve` button, and finally click `Submit review` to formally approve the PR.
```

>   -  If the code changes look good, the approver navigates to the `Reviews changes` section in the GitHub UI. They then enter either `/lgtm` or `/approve` as a message, check the GitHub `Approve` button, and finally click `Submit review` to formally approve the PR.

Reviewer: ```suggestion
  -  If the code changes look good, the approver navigates to the _Reviews changes_ section in the GitHub UI. They check the GitHub _Approve_ button and finally click _Submit review_ to formally approve the PR.
```

This would be accurate for Kubernetes but not for use. We require the GitHub UI approval, but we don't need approvers to type `/lgtm` or `/approve`.

Thus, typing `/lgtm` and `/approve` is NOT required here and should not be recommended.

     - This is required since The Falco Project enforces GitHub reviews alongside the Prow workflow.
     - Furthermore, some repositories may require two (or more) positive GitHub reviews (notably https://github.com/falcosecurity/falco and https://github.com/falcosecurity/libs).

>   - A positive GitHub review implies the PR is approved and looks good, so the approver doesn't have to type `/approve` or `/lgtm` explicitly (however some repositories - notably https://github.com/falcosecurity/test-infra may require an explicit `/approve`).

Reviewer 2: I feel we duplicate info here and above, would there be a way to have it written once in a concise way? Ideas?

>   - A positive GitHub review implies the PR is approved and looks good, so the approver doesn't have to type `/approve` or `/lgtm` explicitly (however some repositories - notably https://github.com/falcosecurity/test-infra may require an explicit `/approve`).

Reviewer 2: ```suggestion

```

Reviewer 2: Maybe just remove as we already said it.

Reviewer: I would keep this as-is, and remove the `/approve` / `/lgtm` in the point above. See my above comment.
